paimon-core/src/main/java/org/apache/paimon/iceberg/manifest/IcebergConversions.java

```diff
@@ -74,8 +77,34 @@ public static ByteBuffer toByteBuffer(DataType type, Object value) {
             case DECIMAL:
                 Decimal decimal = (Decimal) value;
                 return ByteBuffer.wrap((decimal.toUnscaledBytes()));
+            case TIMESTAMP_WITHOUT_TIME_ZONE:
+                TimestampType timestampType = (TimestampType) type;
+                return convertTimestampWithPrecisionToBuffer(
+                        (Timestamp) value, timestampType.getPrecision());
+            case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+                LocalZonedTimestampType localTimestampType = (LocalZonedTimestampType) type;
+                return convertTimestampWithPrecisionToBuffer(
+                        (Timestamp) value, localTimestampType.getPrecision());
+            case TIME_WITHOUT_TIME_ZONE:
+                long microsecondsFromMidnight = (Integer) value / 1_000;
+                return ByteBuffer.allocate(8)
+                        .order(ByteOrder.LITTLE_ENDIAN)
+                        .putLong(0, microsecondsFromMidnight);
             default:
                 throw new UnsupportedOperationException("Cannot serialize type: " + type);
         }
     }
+
+    private static ByteBuffer convertTimestampWithPrecisionToBuffer(
+            Timestamp timestamp, int precision) {
+        long timestampValue;
+        long secondsSinceEpoch = timestamp.toInstant().getEpochSecond();
+        if (precision <= 6) {
+            timestampValue =
+                    secondsSinceEpoch * 1_000_000 + timestamp.getNanoOfMillisecond() / 1_000;
+        } else {
+            timestampValue = secondsSinceEpoch * 1_000_000_000 + timestamp.getNanoOfMillisecond();
+        }
```

Comment:
So you calculate seconds (integer part of one second) and nano of milliseconds (starting from 7th digit after decimal point) here. What about the 1st to the 6th digit after decimal point? Why not use `timestamp.millisecond`?

Reply:
Thank you. I took a look at the source code and reinterpreted it
This is my knowledge blind spot, sorry, repeated mistakes